cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,424 @@
+      return {beta, 1.0};
+    }
+  }
+};
+
+template <typename T>
+class ewma_adjust_denominator_nonull_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_adjust_denominator_nonull_functor(T beta) : beta{beta} {}
+
+  __device__ pair_type<T> operator()(pair_type<T> pairs) { return {beta, 1.0}; }
+};
+
+template <typename T>
+class ewma_noadjust_nonull_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_noadjust_nonull_functor(T beta) : beta{beta} {}
+
+  __device__ pair_type<T> operator()(thrust::tuple<T, size_type> data)
+  {
+    T const beta          = this->beta;
+    T const input         = thrust::get<0>(data);
+    size_type const index = thrust::get<1>(data);
+    if (index == 0) {
+      return {beta, input};
+    } else {
+      return {beta, (1.0 - beta) * input};
+    }
+  }
+};
+
+template <typename T>
+class ewma_noadjust_null_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_noadjust_null_functor(T beta) : beta{beta} {}
+
+  __device__ pair_type<T> operator()(thrust::tuple<T, size_type, bool, size_type> data)
+  {
+    T input           = thrust::get<0>(data);
+    size_type index   = thrust::get<1>(data);
+    bool is_valid     = thrust::get<2>(data);
+    size_type nullcnt = thrust::get<3>(data);
+
+    T beta = this->beta;
+
+    if (index == 0) {
+      return {beta, input};
+    } else {
+      if (is_valid and nullcnt == 0) {
+        // preceeding value is valid, return normal pair
+        return {beta, (1.0 - beta) * input};
+      } else if (is_valid and nullcnt != 0) {
+        // one or more preceeding values is null, adjust by how many
+        T factor = (1.0 - beta) + pow(beta, nullcnt + 1);
+        return {(beta * (pow(beta, nullcnt)) / factor), ((1.0 - beta) * input) / factor};
+      } else {
+        // value is not valid
+        return this->IDENTITY;
+      }
+    }
+  }
+};
+
+/**
+* @brief Return an array whose values y_i are the number of null entries
+* in between the last valid entry of the input and the current index.
+* Example: {1, NULL, 3, 4, NULL, NULL, 7}
+        -> {0, 0     1, 0, 0,    1,    2}
+*/
+rmm::device_uvector<cudf::size_type> null_roll_up(column_view const& input,
+                                                  rmm::cuda_stream_view stream)
+{
+  rmm::device_uvector<cudf::size_type> output(input.size(), stream);
+
+  auto device_view = column_device_view::create(input);
+  auto valid_it    = cudf::detail::make_validity_iterator(*device_view);
+
+  // Invert the null iterator
+  thrust::transform(rmm::exec_policy(stream),
+                    valid_it,
+                    valid_it + input.size(),
+                    output.begin(),
+                    [] __device__(bool valid) -> bool { return 1 - valid; });
+
+  // 0, 1, 0, 1, 1, 0 -> 0, 0, 1, 0, 0, 2
+  thrust::inclusive_scan_by_key(
+    rmm::exec_policy(stream), output.begin(), output.end() - 1, output.begin(), output.begin() + 1);
```

Comment:
Combine these into a scan with inversion applied as a transform on the iterator. However, I note you shift the output by one, which means you leave the first element as its null value (0 or 1). But this means that this first value doesn't influence the scan. Is that intentional?  If it is, my suggestion will need modification.

Also note using `not valid` instead of `1 - valid` to reduce unneeded integer casting.

```suggestion
  // Invert the validity iterator to get nulls
  auto nulls = thrust::make_transform_iterator(valid_it, [] __device__(bool valid) { return not valid; });

  // Scan the nulls 
  // 0, 1, 0, 1, 1, 0 -> 0, 1, 0, 1, 2, 0
  thrust::inclusive_scan_by_key(rmm::exec_policy(stream), nulls, nulls + input.size(), nulls, output.begin());
```

Reply:
I am having a little trouble getting this to compile, but let me address your question about the first value in the mean time. The reason the output is shifted is because I want to be able to line each value up with the number of nulls that came before it. But since this is the first value, there can't be any nulls that come before it, so this number will always be zero, except in the case where the first value is itself null. Since there didn't seem to be a great way of making sense of a sequence that starts with missing values, I had expected there would be UB in that case.

The only other option I could see would be to find the first non null value and execute the existing algorithm from that point forward in the data, leaving the leading nulls as they are - but I can't convince myself that doesn't lead to some conflicting assumptions (don't we assume the first value is, in fact, the first value ever in the `FINITE` case?)

Reply:
Just as an update I was never able to produce correct results using this approach. I am fairly (?) sure that this is the logical equivalent:
```
  auto nulls = thrust::make_transform_iterator(valid_it, [] __device__(bool valid) { return not valid; });
  thrust::inclusive_scan_by_key(rmm::exec_policy(stream), nulls, nulls + input.size() - 1, nulls, output.begin() + 1);
```
However this just doesn't produce the same output sequence as the current implementation. I suspect there's something subtle going on with the iterators that escapes me, I tried several permutations including creating separate iterators representing the same null sequence and so forth, but didn't find success. I am leaving this as-is unless @harrism has any insight on what might be going wrong.

Reply:
We resolved this in the latest. The issue with Brandon's snippet above is that the lambda was accepting and outputting booleans. That's fixed now.